enhancements/sig-policy/59-ansible-everyevent-mode/README.md

```diff
@@ -0,0 +1,189 @@
+
+### Non-Goals
+
+1. Consolidating Ansible job runs when multiple managed clusters are in violation.
+1. Keeping track of every historical `PolicyAutomation` run.
+1. Adding additional `eventHook` options.
+
+## Proposal
+
+To configure Ansible automation when a policy violation occurs, a `PolicyAutomation` object is created which references
+the triggering policy as shown below.
+
+```yaml
+apiVersion: policy.open-cluster-management.io/v1beta1
+kind: PolicyAutomation
+metadata:
+  name: create-ticket
+spec:
+  mode: everyEvent
+  policyRef: enable-etcd-encryption
+  eventHook: noncompliant
+  automationDef:
+    name: Demo Job Template
+    secret: ansible-automation-platform-secret-name
+    extra_vars:
+      sn_severity: 1
+      sn_priority: 1
+```
+
+The proposal is to add a new mode of `everyEvent` that runs an Ansible job on every unique policy violation per managed
+cluster. In order to achieve this, the status of the `PolicyAutomation` must be able to track which managed clusters the
+Ansible job has run for already.
+
+Take for example a scenario where a policy called `enable-etcd-encryption` applies to managed clusters `cluster1` and
+`cluster2`. The user has also defined a `PolicyAutomation` object and set the `mode` to `everyEvent`. Then the
+`enable-etcd-encryption` policy becomes noncompliant (i.e. violated) on `cluster1`. The Governance Policy Propagator
+controller would run the Ansible job from the Hub while passing the `extra_var` of `target_cluster: cluster1`. One
+option is to reuse the `target_clusters` `extra_var` that is used during the `once` mode, however, it is misleading to
+provide an array that will always have a length of 1. The controller would then set the following status on the
+`PolicyAutomation` object. Note that there is currently no status that ever gets set on `PolicyAutomation` objects.
+
+```yaml
+status:
+  clustersWithEvent:
+    - cluster1
+```
+
+This status update would tell the Governance Policy Propagator controller to not run the Ansible job again on `cluster1`
+the next time it notices that the `enable-etcd-encryption` policy is noncompliant on `cluster1` such as during periodic
+reconciling or status updates. This isn't meant to store a complete history of which clusters had an Ansible job run due
+to a policy violation. It is meant to track the clusters **currently** in the state based on the `eventHook` value in
+the `PolicyAutomation` object and the appropriate automation was run for those clusters. Such a history can be found on
+the Ansible Tower/Ansible Automation Platform job history.
+
+Several minutes later, the `enable-etcd-encryption` policy becomes noncompliant on `cluster2`. The Governance Policy
+Propagator controller would run the Ansible job from the Hub while passing the `extra_var` of
+`target_cluster: cluster2`. It would then set the following status on the `PolicyAutomation` object.
```

Comment:
Have you thought about a "mass noncompliance" scenario?  What if something happens (maybe a bad edit to the policy) that causes a currently fully compliant policy to suddenly become NonCompliant across all managed clusters?  If there are a large number of managed clusters, are we creating a large number of automations - 1 for each managed cluster?

Reply:
Maybe a realistic scenario would be using hub templates to push a certificate from the hub to managed clusters.  Now that certificate expires -- causing a resource managed by a policy to become noncompliant everywhere it's deployed, I wonder if you can add some comments to help understand that multiple clusters becoming noncompliant at once are properly handled.

Reply:
@gparvin yes, it'd be one Ansible job per cluster. I don't think it'd overwhelm the Ansible Tower/AAP much more than if we consolidated them, but it could potentially overwhelm the Ansible Operator and create a lot of objects in etcd.

Alternatively, we could continue using `target_clusters: [...]` as the extra var and group them together when we can. I was originally thinking `target_cluster` would be an easier interface to work with, but I'll go back to `target_clusters: [...]` to be safe.